doc/neps/ufunc-overrides.rst

```diff
@@ -153,49 +152,207 @@ tuple in the ``out`` keyword argument.
 
 The function dispatch proceeds as follows:
 
-- If one of the input arguments implements ``__array_ufunc__`` it is
-  executed instead of the Ufunc.
+- If one of the input or output arguments implements
+  ``__array_ufunc__``, it is executed instead of the ufunc.
 
 - If more than one of the input arguments implements ``__array_ufunc__``,
   they are tried in the following order: subclasses before superclasses,
-  otherwise left to right.  The first ``__array_ufunc__`` method returning
-  something else than :obj:`NotImplemented` determines the return value of
-  the Ufunc.
+  inputs before outputs, otherwise left to right.
+
+- The first ``__array_ufunc__`` method returning something else than
+  :obj:`NotImplemented` determines the return value of the Ufunc.
 
 - If all ``__array_ufunc__`` methods of the input arguments return
   :obj:`NotImplemented`, a :exc:`TypeError` is raised.
 
-- If a ``__array_ufunc__`` method raises an error, the error is propagated
-  immediately.
-
-If none of the input arguments has an ``__array_ufunc__`` method, the
-execution falls back on the default ufunc behaviour.
+- If a ``__array_ufunc__`` method raises an error, the error is
+  propagated immediately.
+
+- If none of the input arguments had an ``__array_ufunc__`` method, the
+  execution falls back on the default ufunc behaviour.
+
+In the above, there is one proviso: if a class has an
+``__array_ufunc__`` attribute but it is identical to
+``ndarray.__array_ufunc__``, the attribute is ignored.  This happens for
+instances of `ndarray` and for `ndarray` subclasses that did not
+override their inherited ``__array_ufunc__`` implementation.
+
+
+Type casting hierarchy
+----------------------
+
+Similarly to the Python operator dispatch mechanism, writing ufunc
+dispatch methods requires some discipline in order to achieve
+predictable results.
+
+In particular, it is useful to maintain a clear idea of what types can
+be upcast to others, possibly indirectly (i.e. A->B->C is implemented
+but direct A->C not). Moreover, one should make sure the implementations of
+``__array_ufunc__``, which implicitly define the type casting hierarchy,
+don't contradict this.
+
+It is useful to think of the typecasting hierarchy as a graph (see
+example below) in which, for any given class A, all other classes that
+define ``__array_ufunc__`` must belong to exactly one of three groups
+(making this an directed acyclic graph):
```

Comment:
I would really include the original definition I had:
```
   For any two different types *A*, *B*, the relation "A can handle B" 
   defined as::

       a.__array_ufunc__(..., b, ...) is not NotImplemented

   for instances *a* and *b* of *A* and *B*, defines the
   edges B->A of a graph.

   This also assumes A.__array_ufunc__ either returns
   an instance of A, NotImplemented, or raises an error.
```
otherwise